[NEW] Feature/custom oauth mail field and interpolation for mapped fields

benkroeger opened the pull request:
<!-- INSTRUCTION: Your Pull Request name should start with one of the following tags -->
<!-- [NEW] For new features -->
<!-- [FIX] For bug fixes -->
<!-- [BREAK] For pull requests including breaking changes -->

<!-- INSTRUCTION: Inform the issue number that this PR closes, or remove the line below -->
CLOSES #11647 
CLOSES #15693 
CLOSES #13220

<!-- INSTRUCTION: Link to a https://github.com/RocketChat/docs PR with added/updated documentation or an update to the missing/outdated documentation list, see https://rocket.chat/docs/contributing/documentation/  -->

<!-- INSTRUCTION: Tell us more about your PR with screen shots if you can -->

This PR adds the config option to pick `email` value from a different identity property (e.g. `upn`)

It came to live when trying to use the [O365 Graph user api](https://docs.microsoft.com/en-us/graph/api/user-get) as `identityPath`. That endpoint provides the user's mail address under the `mail` property, which would cause the `email` field in the user's identity to be empty.
Also, since O365 Graph api adds a property `@odata.context` to the response json and mongodb doesn't allow document paths to contain `.` characters, this PR removes all properties that contain a `.` as the very last step in `normalizeIdentity` (happening last in order to allow intermediate steps to still work with those properties; e.g. lookup `@odata....` in the `getUsername` method).

Also, when configuring Azure AD records to contain values like `employeeId`, this could become a powerful combination to populate user records in RocketChat (`mail` as email and `employeeId` as username)

**UPDATE**
73aebfa0578ffe2ba0038e51cef51833a022b230 adds the possibility to use string interpolation in custom field mappings.
this allows to mix and match fragments of existing fields (identity response from provider) into new values (e.g. pick the leading part of email and use as username: `{{/^(.+)@/::email}}`)

Note:
- if you want interpolation, the field mapping must contain text in double curly braces `{{interpolation here}}`
- interpolation can be anything from a simple field mapping `{{email}}` through a nested path `{{foo.bar}}` to a regexp applied to the value from a nested path `{{/^(.+)@/::profile.email}}`
- interpolations can be combined with static strings `{{/^(.+)@/::email}}@rocket.chat`. assuming `email` has the value `foo@bar.com` the result would be `foo@rocket.chat`
- regular expressions in interpolations are enclosed with slashes `/` and separated from the field selector with double colons `::`
- regular expressions must have a single capture group. the value from the first capture group is used for the interpolation value

benkroeger commented on app/custom-oauth/server/custom_oauth_server.js:
```diff
@@ -308,50 +359,56 @@ export class CustomOAuth {
-		if (!avatarUrl) {
-			logger.debug(`Avatar field "${ this.avatarField }" not found in data`, data);
+		if (!value) {
+			throw new Meteor.Error('field_not_found', `Avatar field "${ this.avatarField }" not found in data`, data);
```
this would be a breaking change... should potentially changed back to a simple `debug` log vs throwing an `error`.
let me know. happy to change

benkroeger commented:
@rodrigok @engelgabriel  can someone pls re-trigger the test-with-oplog-mongo-4-0? I didn't change anything related in the last commit (previous commit ran tests successfully). Suspecting a hick-up in circle-ci here.

thanks!

tammer123 commented:
Is there any updates with this?

tammer123 commented:
Updates?

benkroeger commented:
@sampaiodiego since this one is similar to #15743, can you take a look?

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=15690) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=15690) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RocketChat/Rocket.Chat?pullRequest=15690) it.</sub>

benkroeger commented:
nothing? in a month? is anything missing?

tammer123 commented:
Maybe you should @ them?

benkroeger commented:
"them" whom? mentioned a few of the frequent contributors already

On Wed, Dec 18, 2019 at 11:39 AM tammer123 <notifications@github.com> wrote:

> Maybe you should @ them?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RocketChat/Rocket.Chat/pull/15690?email_source=notifications&email_token=AB3L5J33SZUR6MGNVBQDE5LQZH4VPA5CNFSM4JF6FDFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHFWBTA#issuecomment-566976716>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3L5J4ROPJBA32D4RCVOFLQZH4VPANCNFSM4JF6FDFA>
> .
>

benkroeger commented:
@tammer123 so you are a team member (now)?

benkroeger commented:
@sampaiodiego since you authored the PR for the latest release, I assume you're pretty active in the development team. I've been trying to drag some attention to this PR for quite a while now. Can you maybe give a hint on what action I'm missing - or even help me get this thing processed? Any comment is highly appreciated

gabriellsh commented:
Hey @benkroeger! How's it going? Thank you really much for your effort man, your review is on the way!

Meanwhile, could you please reset both History files so they don't show as edited? Just to get this out of the way :).

benkroeger commented:
@gabriellsh is there an average timeframe for a review? I don't want to bother every other day - but the longer this takes, the more often I need to rebase and merge (done two or three times already)

pierre-lehnen-rc commented on app/custom-oauth/server/custom_oauth_server.js:
```diff
@@ -344,50 +394,57 @@ export class CustomOAuth {
+		// remove all properties with names that don't match the RegExp in `IDENTITY_PROPNAME_FILTER`
+		return Object.entries(identity).reduce((result, [name, value]) => {
+			if (IDENTITY_PROPNAME_FILTER.test(name)) {
+				return result;
```
Can you replace the `.` with underscores instead of discarding the attributes? That way we will continue to store the whole data.

benkroeger replied:
well, this might get a little tricky due to nested properties. I believe, the odata protocol allows to have multiple levels of props, each of which can have 'invalid' names. Also, there might be arrays that contain objects with such field names.
if you need any value from odata to build the profile, you can use it through the introduces templating mechanism. these fields are only removed after all the templating happened.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=15690) <br/>All committers have signed the CLA.

benkroeger commented:
@pierre-lehnen-rc @gabriellsh I sense silence again. what's wrong?

jacobsenj commented:
We would need this feature!

rodrigok commented:
@benkroeger sorry for the long delay here on review. Let's make this happen.

I reviewed your can and it seems ok, I didn't test it yet but I'd like to ask you if you could add unit tests for the interpolation method, It's really great and I think we can use it in other places as well, but It's hard to review and understand if everything is ok. Is it possible to create those tests? Let me know if you have any questions on how to create those tests.

benkroeger commented:
@rodrigok unit tests added - for simplicity in testing, I outsourced the newly added methods into a separate file. this also reduces the risk of code conflicts in the main file

rodrigok commented:
@benkroeger great, thanks, I'll review it again soon

rodrigok commented:
@benkroeger I tried this config:
![image](https://user-images.githubusercontent.com/234261/77607481-38f95d80-6ef9-11ea-8cb8-cdd4e02ee0e8.png)

With this value returning on `/me`:
![image](https://user-images.githubusercontent.com/234261/77607501-4f071e00-6ef9-11ea-8d62-402772013734.png)

And got this result:
![image](https://user-images.githubusercontent.com/234261/77607519-60502a80-6ef9-11ea-9b2d-4244e2b94ca3.png)

The email is correct but the name and username are wrong

benkroeger commented:
@rodrigok can not explain where die mismatch comes from. added unit test to process exactly your test data and template expression - receiving the expected results. I also added the normalizers to `template_helpers` and included 'normalizing' for the test data (suspected that some data manupulation might happen there).

is it possible that you ran tests a couple of times, maybe even with different setup or an existing user with the same id?
I can not reproduce what you see.

benkroeger commented:
live test on a running instance also works fine for me (completely fresh meteor instance), using my actual azure ad oauth configuration:

![image](https://user-images.githubusercontent.com/7782055/77643473-b6f04f80-6f5f-11ea-8155-1b3b632fa3c8.png)

![image](https://user-images.githubusercontent.com/7782055/77643490-bd7ec700-6f5f-11ea-8bdd-e68de4dbc6b6.png)

![image](https://user-images.githubusercontent.com/7782055/77643501-c2437b00-6f5f-11ea-94b5-ae17d59ba222.png)

rodrigok commented:
@benkroeger you need to apply this change:
```
From 22db311f707550e480f1d208740ba6b5097facc7 Mon Sep 17 00:00:00 2001
From: Rodrigo Nascimento <rodrigoknascimento@gmail.com>
Date: Thu, 26 Mar 2020 15:25:26 -0300
Subject: [PATCH] Do not process OAuth fields 2 times

---
 app/custom-oauth/server/custom_oauth_server.js | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/app/custom-oauth/server/custom_oauth_server.js b/app/custom-oauth/server/custom_oauth_server.js
index 020da8782..c37e43b62 100644
--- a/app/custom-oauth/server/custom_oauth_server.js
+++ b/app/custom-oauth/server/custom_oauth_server.js
@@ -357,15 +357,15 @@ export class CustomOAuth {
 			}
 
 			if (this.usernameField) {
-				user.username = this.getUsername(user.services[this.name]);
+				user.username = user.services[this.name].username;
 			}
 
 			if (this.emailField) {
-				user.email = this.getEmail(user.services[this.name]);
+				user.email = user.services[this.name].email;
 			}
 
 			if (this.nameField) {
-				user.name = this.getCustomName(user.services[this.name]);
+				user.name = user.services[this.name].name;
 			}
 
 			if (this.mergeRoles) {
-- 
2.21.1 (Apple Git-122.3)

```

I've tried to push to your branch but I was blocked.

rodrigok commented:
The current code is processing the fields 2 times

benkroeger commented:
I've rebased my work on #develop regularly (last today with d223fb1). Is this something you just added or should this have been there before?
Also, this issue would happen even without my proposed changes. Shouldn't it be fixed in a separate PR?

rodrigok commented:
@benkroeger I didn't add this to our code. It wasn't an issue before because wasn't possible to combine values via interpolation. For example, now the `name` property may be a mix of `{{name}} {{anotherField}}` and if this process again, since I'm using the current field inside the interpolation it will get name as the already changed value and will contact with `anotherField` again.

benkroeger commented:
@rodrigok added your suggested modifications (and rebased again)

rodrigok commented:
@benkroeger thanks man, it was a great contribution.

We are now in the feature freeze period during the Release Candidate of version 3.1. So we will merge this in a few days for the development of version 3.2.

luixxiul commented:
@benkroeger Glückwunsch 🥳
